src/plugins/intel_gpu/include/intel_gpu/runtime/memory.hpp

```diff
@@ -258,4 +259,29 @@ inline ov::Tensor make_tensor(layout l, void* memory_pointer) {
     return ov::Tensor(et, l.get_shape(), memory_pointer);
 }
 
+struct MemoryUsageTracker {
+public:
+    MemoryUsageTracker(const engine* engine, float buffers_preallocation_ratio)
+        : _engine(engine)
+        , _buffers_preallocation_ratio(buffers_preallocation_ratio) {
+        static_assert(_max_deque_size >= 2, "[GPU] Deque is supposed to contain at least 2 elements for prediction");
+    }
+
+    std::pair<bool, ov::Shape> predict_preallocated_shape_size(const std::string& id, const ov::Shape& current_shape, bool can_reuse_buffer);
+    bool can_preallocate(size_t desired_buffer_size);
+
+private:
+    enum class math_op { SUM, SUB, MUL };
+    ov::Shape shapes_math(const ov::Shape& shape1, const ov::Shape& shape2, math_op op);
+    void add_shape(const std::string& id, const ov::Shape& shape);
```

Comment:
Does it need to works with shapes at all? Won't buffer byte size be enough for tracking & prediction?

Reply:
Sounds good, but I think shapes themself still may be useful for LLMs cases detection (for example shapes [128, 128] -> [256, 128] -> [192, 256] may be considered as sequential by buffers sizes (16384, 32768, 49152), but this is not a sequence in terms of iterations). So, it looks like we need to collect shapes as well

Reply:
Does it really matter? If you change different dims of the shape on each iteration, but buffer size increase is same on each iter, then why we need to ensure that only specific dim is updated?

Reply:
@vladimir-paramuzov, such a sequence is most likely is some kind of successful combination of circumstances, isn't? I'm not sure we can rely a lot on it without extra conditions, otherwise we run the risk of allocation extra memory, which will never be used, especially taking into account the fact that we don't have mechanism for memory usage reduction. I'd suggest to keep shape usage for LLMs cases as an additional more strict rule for initial implementation...
It would be great to track and detect mentioned above cases as well, but it looks like this requires wider buffer size/shape changes history collection to make sure the trend is upward, and then we don't need to have exactly the same size change between iterations; and also it would be great to have mechanism for reducing memory buffer sizes if the last N iterations do not reach maximum size